### PR TITLE
Allow multiline TextInputs be submittable without blurring

### PR DESCRIPTION
**Pull request by tonyd33**
## Summary

For multiline TextInputs, it's possible to send the submit event when pressing the return key only with `blurOnSubmit`. However, there's currently no way to do so without blurring the input and dismissing the keyboard. This problem is apparent when we want to use a TextInput to span multiple lines but still have it be submittable (but not blurrable), like one might want for a TODO list.

multiline-momentary-blur

## Changelog

[General] [Added] - Add `returnKeyAction` prop to `TextInput` component
[General] [Deprecated] - Remove usages of `blurOnSubmit` in native code and convert `blurOnSubmit` to `returnKeyAction` in the JavaScript conversion layer

## Test Plan

Verified old usages of combinations of `blurOnSubmit` and `multiline` matched previous behavior and that the new `returnKeyAction` prop behaves as expected.

| Android | iOS |
| --- | -- |
| android-returnkeyaction-test | ios-returnkeyaction-test  |

With the changes, the TODO list example from before now looks like this:

multiline-no-momentary-blur

### CODE REVIEW

File: ReactCommon/react/renderer/components/textinput/iostextinput/conversions.h [hunk 1]
```diff
@@ -126,6 +126,30 @@ inline void fromRawValue(
   abort();
 }
 
+inline void fromRawValue(
+    const PropsParserContext &context,
+    const RawValue &value,
+    ReturnKeyAction &result) {
+  auto string = (std::string)value;
+  if (string == "default") {
```
Comment (lunaleaps) [inline ReactCommon/react/renderer/components/textinput/iostextinput/conversions.h:134]: Where do we set `"default"`? Is this set somewhere when the component doesn't pass the prop?

File: ReactCommon/react/renderer/components/textinput/androidtextinput/react/renderer/components/androidtextinput/AndroidTextInputProps.cpp [hunk 2]
```diff
@@ -145,10 +145,10 @@ AndroidTextInputProps::AndroidTextInputProps(
...
           sourceProps.selectTextOnFocus,
           {false})),
-      blurOnSubmit(convertRawProp(context, rawProps,
-          "blurOnSubmit",
-          sourceProps.blurOnSubmit,
-          {false})),
+      returnKeyAction(convertRawProp(context, rawProps,
+           "returnKeyAction",
+          sourceProps.returnKeyAction,
+          {})),
```
[the remark this answers is not in the document]
Comment (tonyd33) [reply ReactCommon/react/renderer/components/textinput/androidtextinput/react/renderer/components/androidtextinput/AndroidTextInputProps.cpp:151]: We're no longer using "blurOnSubmit" in native code as it's being replaced by "returnKeyAction". This conversion is done in TextInput.js

File: ReactCommon/react/renderer/components/textinput/iostextinput/conversions.h (hunk 1 above)
Comment (tonyd33) [reply ReactCommon/react/renderer/components/textinput/iostextinput/conversions.h]: No, it should always be set to one of `"blur"`, `"blurAndSubmit"` or `"newline"`. This is removed now

**Comment by makovkastar** [comment]
Posting some of @yungsters's feedback on this PR: 

After thinking about this more, I don't think "return key" is the right name for this prop.

The button on mobile keyboards is not necessarily a "return key" like it is on web and desktop platforms. Neither "return key" nor "submit key" are great because both imply one of the behaviors: 
- "return key" implies a default behavior of inserting a new line.
- "submit key" implies a default behavior of invoking onSubmitEditing.

That said, I think it would still be better to use terminology that already exists in the interface. So… maybe submitBehavior or submitAction instead of returnKeyAction?

**Comment by tonyd33** [comment]
> Posting some of @yungsters's feedback on this PR:
> 
> After thinking about this more, I don't think "return key" is the right name for this prop.
> 
> The button on mobile keyboards is not necessarily a "return key" like it is on web and desktop platforms. Neither "return key" nor "submit key" are great because both imply one of the behaviors:
> 
> * "return key" implies a default behavior of inserting a new line.
> * "submit key" implies a default behavior of invoking onSubmitEditing.
> 
> That said, I think it would still be better to use terminology that already exists in the interface. So… maybe submitBehavior or submitAction instead of returnKeyAction?

Yeah. All of returnKeyAction, submitBehavior, and submitAction face a similar problem but the latter two probably address it better than returnKeyAction. This has been changed

**Comment by jamninetyfive** [comment]
@tonyd33  any update? this pr sloved my problem , use multiline and onSubmitEditing same time and when press send don't track newline.

**Comment by hsource** [comment]
@makovkastar - we'd love to get this merged - is there anything else you need from @tonyd33 on this?

**Pull request closed without merging**